File: forC++/92.反转链表-ii.cpp
/*
 * @lc app=leetcode.cn id=92 lang=cpp
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution
{
public:
    ListNode *reverseBetween(ListNode *head, int left, int right)
    {
        if(left==1){//
           return revertList(head,right);
        }
        head->next = reverseBetween(head->next,left-1,right-1);
        return head;
    }
    ListNode *currListNode=nullptr;
    ListNode *revertList(ListNode *head,int index){
        if(index==1){//
            currListNode=head->next;//
            return head;//
        }
        ListNode *listNode=revertList(head->next,index-1);
        head->next->next=head;
        head->next=currListNode;
        return listNode;//
    }
};
// @lc code=end

File: CodeTopGo/92.反转链表-ii.go
/*
 * @lc app=leetcode.cn id=92 lang=golang
 *
 * [92] 反转链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
	hair := &ListNode{}
	hair.Next = head
	first := hair
	second := head
	for i := 0; i < left-1; i++ {
		first = first.Next
	}
	second = first.Next
	for j := left; j <= right; j++ {
		second = second.Next
	}

	first.Next = reverseList(first.Next, second)
	return hair.Next
}

func reverseList(start *ListNode, end *ListNode) *ListNode {
	prev := end
	curr := start
	for curr != end {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}
	return prev
}

// @lc code=end

